### PR TITLE
gadget: refactor StructureEncryption to have a concrete type instead of map

### PR DESCRIPTION
**Pull request by anonymouse64**
Followup to https://github.com/snapcore/snapd/pull/11354, also introduces some stuff to gadgettest, which we use in https://github.com/snapcore/snapd/pull/11374.

### CODE REVIEW

**Review by stolowski — APPROVED** [review]
LGTM, thank you

File: gadget/gadget.go [hunk 1]
```diff
@@ -252,10 +252,45 @@ type DiskVolumeDeviceTraits struct {
...
+	// recognized.
+	Method DiskEncryptionMethod `json:"method"`
+
+	// unknownKeys is used to log messages about unknown, unrecognized keys that
+	// we may encounter and may be used in the future
+	unknownKeys map[string]string
+}
+
+func (s *StructureEncryptionParameters) UnmarshalJSON(b []byte) error {
+	m := map[string]string{}
```
Comment (mardy) [inline gadget/gadget.go:276]: I would happily sacrify the logging of unknown keys and directly unmarshal into our struct, for simplicity. But that's MHO :-)

**Review by mardy — APPROVED** [review]
LGTM! Minor nitpicks/personal preferences :-)

File: gadget/gadget.go [hunk 2]
```diff
@@ -252,10 +252,45 @@ type DiskVolumeDeviceTraits struct {
 	Structure []DiskStructureDeviceTraits `json:"structure"`
 
 	// StructureEncryption is the set of partitions that are encrypted on the
-	// volume - this should only ever be ubuntu-data or ubuntu-save for now, but
-	// the map will indicate the name of the structure as the key and the type
-	// of encryption (currently only "LUKS") as the value in the map.
-	StructureEncryption map[string]map[string]string `json:"structure-encryption"`
+	// volume - this should only ever have ubuntu-data or ubuntu-save for now
+	// keys in the map. The value indicates parameters of the encryption present
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply gadget/gadget.go:256]: fixed in https://github.com/snapcore/snapd/pull/11374, thanks for spotting

**Pull request merged**